File: packages/sdk-node/src/template.ts
import type { Message, PromptTemplateRef } from "./types.js";

/**
 * Infer a prompt template from the system message when the caller didn't declare one.
 * Heuristic: the system prompt is the template; interpolated values are usually in user turns.
 * Callers get far better version tracking by passing { promptTemplate: { name, raw } } explicitly.
 */
export function inferTemplate(messages: Message[], system?: unknown): PromptTemplateRef | undefined {
  const sys = typeof system === "string" ? system : contentToText(messages.find((m) => m.role === "system")?.content);
  if (!sys) return undefined;
  return { name: `system:${fnv1a(sys).slice(0, 8)}`, raw: sys };
}

export function contentToText(content: unknown): string {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content.map((b) => (typeof b === "string" ? b : b && typeof b === "object" && "text" in b ? String((b as { text: unknown }).text) : "")).join("");
  }
  return "";
}

function fnv1a(str: string): string {
  let h = 0x811c9dc5;
  for (let i = 0; i < str.length; i++) { h ^= str.charCodeAt(i); h = Math.imul(h, 0x01000193) >>> 0; }
  return h.toString(16).padStart(8, "0");
}

File: packages/sdk-python/regressa/template.py
from __future__ import annotations

import hashlib
from typing import Any


def content_to_text(content: Any) -> str:
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts = []
        for block in content:
            if isinstance(block, str):
                parts.append(block)
            elif isinstance(block, dict) and "text" in block:
                parts.append(str(block["text"]))
            elif hasattr(block, "text"):
                parts.append(str(block.text))
        return "".join(parts)
    return ""


def infer_template(messages: list[dict[str, Any]], system: Any = None) -> dict[str, str] | None:
    """Fallback template detection: the system prompt is treated as the template.

    Pass prompt_template={"name": ..., "raw": ...} explicitly for accurate version tracking.
    """
    sys_text = system if isinstance(system, str) else None
    if sys_text is None:
        for m in messages:
            if m.get("role") == "system":
                sys_text = content_to_text(m.get("content"))
                break
    if not sys_text:
        return None
    digest = hashlib.sha1(sys_text.encode("utf-8")).hexdigest()[:8]
    return {"name": f"system:{digest}", "raw": sys_text}
